Add documentation for Serving Encryption (#5804)
knative/docs: docs/serving/encryption/configure-certmanager-integration.md

# Configuring Knative cert-manager integration

> Knative Serving relies on a bridging-component to use cert-manager for automated certificate provisioning. 

Reviewer: s/bridging-component/bridging component

The following must be installed on your Knative cluster:

- [Knative Serving](../../install/yaml-install/serving/install-serving-with-yaml.md).
- [`cert-manager`](../../install/installing-cert-manager.md) version `1.0.0` or higher.

!!! warning
    Make sure you have installed cert-manager. Otherwise, the Serving controller will not start up correctly.

## Issuer configuration

The Knative cert-manager integration defines three references to [cert-manager issuers](https://cert-manager.io/docs/configuration/issuers/) to configure different CAs
for the [three Knative Serving encryption features](./encryption-overview.md):

* `issuerRef`: issuer for external-domain certificates used for ingress.
* `clusterLocalIssuerRef`: issuer for cluster-local-domain certificates used for ingress.
* `systemInternalIssuerRef`: issuer for certificates for system-internal-tls certificates used by Knative internal components.

The following example uses a self-signed `ClusterIssuer` and the Knative cert-manager integration references that `ClusterIssuer` for all three configurations.
As this **should not be used in production** (and does not support rotating the CA without downtime),

> you should think about which CA should be used for each use-case and how trust will be distributed to the clients calling the encrypted services. 

Reviewer: s/use-case/use case

| system-internal-tls      | Self-signed Cert-Manager `ClusterIssuer`   | The CA will be populated by cert-manager. DevOps team will use [trust-manager](https://cert-manager.io/docs/trust/trust-manager/) to distribute the CA to Knative system components. |

### Issuer selection

In general, you can refer to the [cert-manager documentation](https://cert-manager.io/docs/configuration/acme/#creating-a-basic-acme-issuer). There are examples available for:

* [CA based on a K8s secret](https://cert-manager.io/docs/configuration/ca/)
* [HTTP-01 challenges, e.g. Let's encrypt](https://cert-manager.io/docs/configuration/acme/#creating-a-basic-acme-issuer)
* [DNS-01 challenges](https://cert-manager.io/docs/configuration/acme/dns01/)
* [Self-signed issuers](https://cert-manager.io/docs/configuration/selfsigned/)

!!! important
    Please note, that not all issuer types work for each Knative feature.

`cluster-local-domain-tls` needs to be able to sign certificates for cluster-local domains like `myapp.<namespace>`, `myapp.<namespace>.svc` and `myapp.<namespace>.svc.cluster.local`. 

> As the CA usually is not within the cluster, verification via ACME protocol (DNS01/HTTP01) is not possible. You can use an issuer that allows creating the these certificates (e.g. CA issuer).

Reviewer 2: ```suggestion
The CA is usually outside the cluster, so verification via ACME protocol (DNS01/HTTP01) is impossible. You can use an issuer that allows the creation of these certificates (e.g., a CA issuer).
```

`system-internal-tls` needs to be able to sign specific SANs that Knative validates for. The defined set of SANs is:

* `kn-routing`
* `kn-user-<namespace>` (<namespace> is each namespace where Knative Services are/will be created)
* `data-plane.knative.dev`

> As this is also not possible via ACME protocol (DNS01/HTTP01), you need to configure an issuer that allows creating the these certificates (e.g. CA issuer).

Reviewer 2: ```suggestion
As this is also not possible via the ACME protocol (DNS01/HTTP01), you need to configure an issuer that allows the creation of these certificates (e.g., a CA issuer).
```

trust needs to be distributed:

* **Cluster external client (Browser and/or other application)**: this is considered out of scope of Knative.
* **Cluster internal client (e.g. Knative or Vanilla K8s workload)**: see below.
* **Knative system components (e.g. Activator, Queue-Proxy, Ingress-Controller)**: see below.

### Trusting for cluster internal clients (e.g. Knative or Vanilla K8s workload)

As Knative does not control all workload and the settings are highly dependent on your runtime and/or language, this is out of scope of Knative.
But here a few points to consider, as there are several ways on how to provide CAs to your application:

* Adding the CA bundle to a Container image on build-time (be aware that this complicates CA rotation, you basically need to rebuild every application)
* Mounting a CA bundle to the filesystem (e.g. from a `Secret` or `ConfigMap`)
* Reading it from environment variable
* Accessing it from a `Secret`/`ConfigMap` via K8s API

If reloading certificates without downtime is important for your client, the workload must either watch changes on the K8s resource (Secret/ConfigMap) or watch the filesystem.

> If the workload is watching the filesystem, it is important to note that using `ionotify` to catch changing Secrets/ConfigMaps is not very reliable on K8s. Tests have shown that it is more reliable to regularly poll and check the certificate on the filesystem for changes.

Reviewer: We will need to have an example on how to do the rotation. I think a minimum example should include manual rotation based on workload restarting (as others have).
